### ISSUE TITLE
device twin simple sample - retrieving version field - on desired property

### ISSUE THREAD

**Issue by kiranpradeep** [Azure/azure-iot-sdk-c#228]
Device twin simple serializer sample([1](https://github.com/Azure/azure-iot-sdk-c/blob/master/serializer/samples/devicetwin_simplesample/devicetwin_simplesample.c)), doesn't demonstrate how to retrieve `version` field from a device twin. On a reconnect, that is the only way to know if the field has been updated([2](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins)).

We are able to retrieve desired property as in the sample code. Is it possible to give hints on retrieving the `version` field in device twin?

**Comment by ewertons** [comment]
Hi @kiranpradeep ,

The only way to obtain the desired properties $version is retrieving the complete Twin json content.
That happens as of now only on the first update you get after calling IoTHubClient_LL_SetDeviceTwinCallback.

The reason why you get the full Twin update after every reconnection is because internally the SDK re-subscribes to Twin after reconnecting, getting the complete json body.

We are working on a API to expose a way for customers to retrieve the full Twin content (with desired and reported properties and their respective versions) on demand.
We believe that should provide a solution for your question.

Could you confirm if that would address your issue?
If you have any additional points to bring to the proposed idea above, please feel free to share them.

**Comment by kiranpradeep** [comment]
@ewertons A full JSON body will force the lib user to parse it just to get the version field. That is not easy on the lib user. If that is the case, you could have always forced the lib user to parse the JSON instead of having easier macros like `WITH_REPORTED_PROPERTY` or `WITH_DESIRED_PROPERTY`. Why does not provide similar macros for version fields or anything easier? 

The version field is very important for any lib user. At the moment, the lib user (on a device) doesn't have any idea if the twin was updated while it was offline. Only was for device client to check if the twin got updated is by looking the version field and comparing it last version field it had received.

Consider this case: Device has a Modbus client and server-side team put the Modbus configuration file (as a URL) in the device twin. The device always retrieves the configuration URL from the twin. On a reconnect, the device doesn't know if the URL is updated or not. It is blind and is always forced to re-download the Modbus register configuration file. The only way to know if the twin is _really_ updated is the version field.

**Comment by ewertons** [comment]
@kiranpradeep , we see your point.
You are interested in using the Serializer client, and have the version property exposed.
If you confirm that is the expectation, I will add a task to our backlog to explore exposing that information on the serializer implementation.

For the scenario you are mentioning, after a device reconnects it does request a full Twin from the service again. It would be expected that the serializer would invoke the model callbacks if any changes happened to the Twin. Are you saying you are experiencing a different behavior?

**Comment by kiranpradeep** [comment]
@ewertons Regarding the first paragraph of your reply, Yes. I like to have version property exposed by Serializer client. 

Regarding the second paragraph of your reply, maybe I misunderstood your reply. But your reply is not appearing in sync with the current behavior of serializer( you mentioned invoking callback if any changes happened which is not true ). On a reconnect, currently, serializer invokes all model callbacks **irrespective** of changes to twin. I understand that IoT Hub does not preserve desired properties update notifications for disconnected devices and so I am fine with the current behavior of serializer. 

My problem is that a reconnected client doesn't have a way to identify if the twin was updated or not. The version field is my only way to find if the twin got updated while I was disconnected. So I requested to expose version field ( how to expose it, should be decided by SDK developers ).

On a second thought, I could also be wrong in my request here. Since SDK developers didn't expose the version field, was it for Azure internal usage? Did SDK developers felt, lib users should never need to identify if a twin got updated during disconnection?
